libsolidity/analysis/NameAndTypeResolver.cpp

```diff
@@ -213,7 +213,14 @@ std::vector<Declaration const*> NameAndTypeResolver::pathFromCurrentScopeWithAll
 
 		pathDeclarations.push_back(candidates.front());
 
-		candidates = m_scopes.at(candidates.front())->resolveName(_path[i]);
+		candidates = m_scopes.at(candidates.front())->resolveName(
+			_path[i],
+			ResolvingSettings{
+				false /* recursive */,
+				true /* alsoInvisible */,
```

Comment:
This implies that invisible declarations will be included in the resolving results. What does that mean?

Reply:
For some reason I don't fully understand, library external functions are registered as "invisible". So, in order to find them here, we need to turn on that setting.

Reply:
For this it'd be important to clarify why exactly it's necessary and double-check the consequences - it does look a bit suspicious.

Reply:
Yeah, I was expecting to get failures after enabling `alsoInvisible`, but didn't get it. I am going to spend more time looking into that.

Reply:
External stuff is generally marked as invisible in contracts:
https://github.com/ethereum/solidity/blob/229fcc9fce5839f65dfaef2d7ddc55bac432fc81/libsolidity/ast/AST.h#L257
https://github.com/ethereum/solidity/blob/229fcc9fce5839f65dfaef2d7ddc55bac432fc81/libsolidity/analysis/NameAndTypeResolver.cpp#L509-L511

I'm not sure why though. I suspected that this is because it's only supposed to be available via qualified access but we have an identifier path here so this processes both qualified and unqualified names. Also it would make `onlyVisibleAsUnqualifiedNames` redundant.

In any case I suspect that this change will have unintended consequences, depending on what stuff gets marked as invisible. I don't know full rules, but looks like some variables get hidden depending on the scope:
https://github.com/ethereum/solidity/blob/229fcc9fce5839f65dfaef2d7ddc55bac432fc81/libsolidity/analysis/NameAndTypeResolver.cpp#L539

Also `this` and `super`.

We could make external stuff visible but I'm pretty sure it will come with even more unintended consequences. If we can't do that, maybe we have to have separate resolution rules specifically for `using for`.

Reply:
I think that the best solution would be to have the separate resolution for `using for`.  
If `registerDeclaration` is changed to make external library functions visible, then they would be allowed also inside the library itself, which then would require a fix be made in order to prevent that. I guess there would be other problems and unexpected situations.